netroute_bsd.go

```diff
@@ -13,137 +13,377 @@
+	}
+
+	return msg, nil
+}
+
+func (r *bsdRouter) getSeq() int {
+	return int(r.seq.Add(1))
+}
+
+// getRouteMsgReply takes an RTM_GET RouteMessage and returns reply (RouteMessage)
+func getRouteMsgReply(msg *route.RouteMessage) (*route.RouteMessage, error) {
+	if msg.Type != syscall.RTM_GET {
+		return nil, errors.New("message type is not RTM_GET")
+	}
+
+	fd, err := getRouteFD()
 	if err != nil {
 		return nil, err
 	}
-	for _, iface := range ifaces {
-		rtr.ifaces[iface.Index] = iface
-		var addrs ipAddrs
-		ifaceAddrs, err := iface.Addrs()
-		if err != nil {
-			return nil, err
-		}
-		for _, addr := range ifaceAddrs {
-			if inet, ok := addr.(*net.IPNet); ok {
-				// Go has a nasty habit of giving you IPv4s as ::ffff:1.2.3.4 instead of 1.2.3.4.
-				// We want to use mapped v4 addresses as v4 preferred addresses, never as v6
-				// preferred addresses.
-				if v4 := inet.IP.To4(); v4 != nil {
-					if addrs.v4 == nil {
-						addrs.v4 = v4
-					}
-				} else if addrs.v6 == nil {
-					addrs.v6 = inet.IP
+	defer unix.Close(fd)
+
+	err = writeMsg(fd, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the response message
+	return readMsg(msg.ID, msg.Seq, fd)
+}
+
+// getRouteFD opens a routing socket.
+// Returns the file descriptor for the socket or an error if the socket cannot be created.
+func getRouteFD() (int, error) {
+	fd, err := unix.Socket(unix.AF_ROUTE, unix.SOCK_RAW, unix.AF_UNSPEC)
+	if err != nil {
+		scope := "route socket - open"
+		return -1, annotateUnixError(err, scope)
+	}
+
+	return fd, nil
+}
+
+// writeMsg marshals and writes a RouteMessage to the specified file descriptor.
+// Returns an error if marshaling fails or if the write operation fails.
+func writeMsg(fd int, msg *route.RouteMessage) error {
+	buf, err := msg.Marshal()
+	if err != nil {
+		return fmt.Errorf("invalid route message given: %s", err)
+	}
+
+	if _, err = unix.Write(fd, buf); err != nil {
+		scope := "route socket - write"
+		return annotateUnixError(err, scope)
+	}
+
+	return nil
+}
+
+// readMsg reads a response from the routing socket and returns the matching RouteMessage.
+// It spawns a goroutine to read from the socket and filters messages based on the provided ID and sequence number.
+// The function includes a 10-second timeout to prevent indefinite blocking.
+func readMsg(id uintptr, seq, fd int) (*route.RouteMessage, error) {
+	type readResult struct {
+		msg *route.RouteMessage
+		err error
+	}
+
+	// Create a channel to receive the result
+	resultCh := make(chan readResult, 1)
+
+	go func() {
+		var rb [2 << 10]byte
+		for {
+			n, err := unix.Read(fd, rb[:])
+			if err != nil {
+				scope := "route socket - read"
+				resultCh <- readResult{
+					nil,
+					annotateUnixError(err, scope),
 				}
+				return
 			}
+
+			// Parse response messages
+			msgs, err := route.ParseRIB(route.RIBTypeRoute, rb[:n])
+			if err != nil {
+				resultCh <- readResult{
+					nil,
+					fmt.Errorf("failed to parse messages: %s", err),
+				}
```

Comment:
This a style thing (so you can ignore it), but I'd suggest using keyed values for each instantiation of this type.

I.e.
```go
resultCh <- readResult{
	err: annotateUnixError(err, scope),
}
```

Here it probably doesn't matter, but generally speaking, this won't break if the typedef members ever shift around.

Reply:
Can do, and yeah probably just good practice